apps/core/actions.py

```diff
@@ -0,0 +1,155 @@
+    2. apply_table(table): Apply the action to table and return new snapshot
+    2. apply_row(row_dict): Apply the action to row dict and return new dict
+    """
+
+    NAME = ""
+    PARAM_TYPES = []
+
+    def __init__(self, params, table: Table):
+        self.is_valid, self.error = self.validate(params, table)
+        self.params = params
+        self.table = table
+
+    def base_validate(self, params: list, _: Table) -> Validation:
+        if len(params) != len(self.PARAM_TYPES):
+            return False, "Invalid number of parameters."
+
+        for i, (param, typ) in enumerate(zip(params, self.PARAM_TYPES)):
+            if not isinstance(param, typ):
+                return (
+                    False,
+                    f"Invalid type {type(param)} for parameter {i}. Expected {typ}.",
+                )
+
+        return True, ""
+
+    def validate_column(self, params: list, table: Table) -> Validation:
+        if len(params) == 0:
+            return True, None
+
+        last_snapshot: Optional[Snapshot] = table.last_snapshot
+        if last_snapshot is None:
+            return False, "No snapshot for table"
+        col_id, *_ = params
+        column_ids = [col["key"] for col in last_snapshot.data_columns or []]
+        if col_id is not None and col_id not in column_ids:
+            return False, "Invalid column id"
+        return True, None
+
+    def validate(self, params: list, table: Table) -> Validation:
+        """
+        Validate the params along with table. The first element of params list is
+        the column name or None which indicates action is applied to all columns.
+        """
+        is_valid, err = self.base_validate(params, table)
+        if not is_valid:
+            return is_valid, err
+
+        return self.validate_column(params, table)
+
+    def apply_table(self, table: Table):
+        """Apply the action to whole table. Results in new set of table rows/columns."""
+        raise MethodNotImplemented
+
+    def apply_row(self, row: dict):
+        """Apply the action to single table row. Results in new set of table row."""
+        raise MethodNotImplemented
+
+
+def get_action_class(name: str) -> Optional[Type]:
+    return __ACTIONS.get(name)
+
+
+def parse_raw_action(raw_action: list, table: Table) -> Optional[BaseAction]:
+    if not raw_action:
+        return None
+    action_name, *params = raw_action
+    Action = get_action_class(action_name)
+    return Action and Action(params, table)
+
+
+@register_action
+class CastColumn(BaseAction):
+    NAME = "cast_column"
+    PARAM_TYPES = [str, str]  # [col_id, type]
+
+    def validate(self, params: list, table: Table) -> Validation:
+        is_valid, err = super().validate(params, table)
+        if not is_valid:
+            return is_valid, err
+        _, target_type = params
+        if target_type not in [
+            ColumnTypes.INTEGER,
+            ColumnTypes.STRING,
```

Comment:
Lets store types in different var.

Reply:
I think I got your question correct. Do you mean storing something like `TARGET_TYPES = [ColumnTypes.INTEGER, ColumnTypes.STRING]`?

In future we'll be using all the enumerations in `ColumnTypes` so this should not be a problem; now, we are just supporting casting into string and int.